source/jcli/cli.d

```diff
@@ -2,327 +2,1929 @@ module jcli.cli;
 
 import jcli;
 
-import std.algorithm;
 import std.stdio : writefln, writeln;
-import std.array;
+import std.meta;
 
-// Needs a complete rework.
-final class CommandLineInterface(Modules...)
+struct ExecuteCommandResult
 {
-    alias Tokenizer     = ArgTokenizer!(string[]);
-    alias bindArgument  = bindArgumentAcrossModules!Modules;
+    int exitCode;
+    /// is null if no exception occured
+    Exception exception;
+}
+
+ExecuteCommandResult executeCommand(T)(auto ref scope T command)
+{
+    try
+    {
+        static if (is(typeof(command.onExecute()) : int))
+        {
+            return typeof(return)(command.onExecute(), null);
+        }
+        else static if (is(typeof(T.onExecute()) == void))
+        {
+            command.onExecute();
+            return typeof(return)(0, null);
+        }
+        else
+        {
+            return typeof(return)(0, null);
+        }
+    }
+    catch (Exception exc)
+    {
+        return typeof(return)(0, exc);
```

Comment:
Return 0 on error?

Reply:
It returns the exception, the user should check that first. I didn't wrap it in a getter, because so far it's only been used internally and extra fluff for data objects is just annoying to me. You can do a SumType here too to ne able to match over the result